8375 ClientService(prepareConnection=...)

twm opened the pull request:
A continuation of the work @dreid began in #90.

## Contributor Checklist:

* [x] There is an associated ticket in Trac. https://twistedmatrix.com/trac/ticket/8375
* [x] There is a newsfragment in src/twisted/newsfragments/ (see: https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles )
* [x] I have updated the automated tests.

Closes #90.

dreid committed:
Add onNewConnection that is called when a new connection is created.

dreid committed:
Merge branch 'trunk' into ClientService-onNewConnection-8375

twm committed:
Merge branch 'trunk' into 8375-clientservice-onnewconnection

Conflicts:
	src/twisted/application/internet.py
	src/twisted/application/test/test_internet.py

twm committed:
Fix test indentation

twm committed:
4 of 5 tests agree: it works!

twm committed:
5 of 5 tests agree: it works!

twm committed:
Add newsfragment

twm committed:
Remove unused variable

twm committed:
Appease txchecker

adiroiban commented on src/twisted/application/internet.py:
```diff
@@ -1029,13 +1058,18 @@ def __init__(self, endpoint, factory, retryPolicy=None, clock=None):
+
+        @param onNewConnection: A single argument L{callable} that may return a
+            Deferred. It will be called with the L{protocol
+            <interfaces.IProtocol>} after a new connection is made.
```
~should this callable return something~ ?

To rephrase the previous question.

Maybe : `A single argument L{callable} that may return a Deferred or None.`

adiroiban commented on src/twisted/application/internet.py:
```diff
@@ -1029,13 +1058,18 @@ def __init__(self, endpoint, factory, retryPolicy=None, clock=None):
         self._machine = _ClientMachine(
             endpoint, factory, retryPolicy, clock,
-            log=self._log,
+            onNewConnection, log=self._log,
```
Maybe also call here using keyword argument?

adiroiban commented on src/twisted/newsfragments/8375.feature:
```diff
@@ -0,0 +1 @@
+twisted.application.internet.ClientService now accepts a function to initialize a connection as the onNewConnection argument
```
Missing a full stop.
Is the function used to initialize the connection or is called when the connection is initialized?

twm committed:
Review feedback

adiroiban commented on src/twisted/newsfragments/8375.feature:
```diff
@@ -0,0 +1 @@
+twisted.application.internet.ClientService now accepts a function which is used to initialize a new connection as the onNewConnection argument.
```
`is used to initialize a new connection` ... maybe 

twisted.application.internet.ClientService now accepts as the `onNewConnection` argument a function which is called when a new connection is made.

adiroiban replied:
`is used to initialize a new connection` ... I don't know if I get what it takes to initialize a new connection?

By reading only this I can understand that the function is in charge of creating/making that connection.

Maybe:

twisted.application.internet.ClientService now has the onNewConnection argument which accepts a function called when a connection is made.

adiroiban commented on src/twisted/application/internet.py:
```diff
@@ -1029,13 +1058,20 @@ def __init__(self, endpoint, factory, retryPolicy=None, clock=None):
+            <interfaces.IProtocol>} after a new connection is made. Should it
+            raise an exception or the L{Deferred} fail, the connection attempt
+            is treated as a failure.
+        @type onNewConnection: L{callable}
```
Why is this "onNewConnection" and not "onConnectionMade" ?

What does "new" mean?

I think that the docstring should be explicit that it is called for each connection (including calling it multiple times) on reconnection after a timeout.... or maybe just call onConnectionMade and this should be straightforward :)

And it would help to have a test in which we check whether it is called or not on reconnection.

adiroiban commented on src/twisted/application/test/test_internet.py:
```diff
@@ -1031,3 +1031,78 @@ def test_stopServiceOnStoppedService(self):
+    def test_onNewConnectionCalledWhenServiceStarts(self):
+        """
+        The C{onNewConnection} callable is called when the
+        L{ClientService.startService} is called.
```
Is this true?
I was expecting onNewConnection to be called only when a connection is done.

If you call ClientService.startService and the connection fails, onNewConnection is not called.

So maybe

onNewConnection is called after ClientService.startService once the connection is made.

adiroiban commented on src/twisted/application/test/test_internet.py:
```diff
@@ -1031,3 +1031,78 @@ def test_stopServiceOnStoppedService(self):
+        def onNewConnection(_proto):
+            newConnections[0] += 1
+
+        cq, service = self.makeReconnector(onNewConnection=onNewConnection)
```
Since the docstring talks about startService,
either use

```
cq, service = self.makeReconnector(onNewConnection=onNewConnection, startService=True)
```

or better, explicit about  what is going on here

```
clock = Clock()
cq, service = self.makeReconnector(
    onNewConnection=onNewConnection,
    startService=False,
    fireImmediatly=False,
    clock=clock)

service.startService()

# The initial connection fails.
cq.connectQueue[0].errback(Exception('failed-connection'))

# Not yet called.
self.assertEqual(0, newConnections[0])

# The connection is made on the retry.
clock.advance(AT_LEAST_ONE_ATTEMPT)
cq.connectQueue[1].callback(None)

self.assertEqual(1, newConnections[0])
```

adiroiban commented on src/twisted/application/test/test_internet.py:
```diff
@@ -1031,3 +1031,78 @@ def test_stopServiceOnStoppedService(self):
+        """
+        newConnections = [0]
+        def onNewConnection(_proto):
+            newConnections[0] += 1
```
It would also help to have a test which checks that whatever onNewConnection returns, in the end, its result is ignored and you get the original protocol

```
newConnectionCalls = []
def onNewConnection(protocol):
    newConnectionCalls.append(protocol)
    return object()

then in the test

self.assertEqual(1, len(newConnectionCalls))
self.assertIs(newConnectionCalls[0], cq.constructedProtocols[0])
```

adiroiban reviewed (dismissed):
Thanks for working on this.

Please check my inline comments.

The  summary:

* onNewConnection vs onConnectionMade
* Have an integration or functional test to also check the condition in which it is not called (ex on connection failure) :)
* Document and test that the result from onNewConnection is ignored an if it want to make a change, it should trigger side effects

twm committed:
Merge branch 'trunk' into 8375-clientservice-onnewconnection

twm committed:
Rename onNewConnection → prepareConnection

Review didn't like the "onNewConnection" name. As "on" prefixes aren't
really used in Twisted, I changed it to "prepareConnection" which is
a bit more suggestive of its purpose.

Additionally, expand the docstring to explain how the callback affects
the retry state machine in more detail. Update the newsfile too.

twm committed:
Stub out new tests

twm committed:
Flesh out tests

twm committed:
Correct style issues

twm committed:
Fix epytext reference issues

twm committed:
Satisfy txchecker

twm commented:
Thank you for the review, @adiroiban! I think that I have addressed all of the feedback:

* I renamed `onNewConnection` to `prepareConnection`. This is more suggestive of the intended purpose of the callback. Dropping the "on" prefix seems more in keeping with the naming of things in Twisted in general, and without the prefix there is no implication that this is merely an observer. The suggestion of "onConnectionMade" would seem to imply the same semantics as IProtocol.connectionMade, but the semantics of this callback are quite different (in particular it returns a Deferred).
* There is enough flexibility inherent in the API design here that I don't think that we can come up with a succinct name which conveys the full semantics of the parameter. We'll have to make up for this with documentation, so I expanded the docstring of the parameter to better describe the ways it interacts with the reconnection loop and clarify that it happens before a `whenConnected()` Deferred resolves.
* I added several test cases suggested by @adiroiban.

The build is failing on the RHEL 7 buildbot worker with a failure to download the appdirs distribution. There are similar failures on this worker for other PRs for different distributions (e.g., #1003 is failing to install pyasn1-modules), so I am marking this for review despite the failure.

twm committed:
Merge branch 'trunk' into 8375-clientservice-onnewconnection

glyph commented on src/twisted/application/internet.py:
```diff
@@ -1029,13 +1058,31 @@ def __init__(self, endpoint, factory, retryPolicy=None, clock=None):
             reactor) will be restored when deserialized.
         @type clock: L{IReactorTime}
+
+        @param prepareConnection: A single argument L{callable} that may return
```
This is the one blocker:  the `prepareConnection` parameter needs a "since" declaration explaining when it's available since.  See https://github.com/twisted/incremental#updating for an explanation about how to phrase this so it will be picked up by the release machinery.

glyph commented on src/twisted/application/internet.py:
```diff
@@ -1029,13 +1058,31 @@ def __init__(self, endpoint, factory, retryPolicy=None, clock=None):
         @type clock: L{IReactorTime}
+
+        @param prepareConnection: A single argument L{callable} that may return
+            a L{Deferred}. It will be called once with the L{protocol
```
Some non-blocking, but useful stuff:

- What should this `Deferred` fire with?
- If it can be anything (which is implied) will its callback type be preserved?
- Will errors be handled?  Logged?

glyph commented on src/twisted/application/internet.py:
```diff
@@ -1029,13 +1058,31 @@ def __init__(self, endpoint, factory, retryPolicy=None, clock=None):
+            authenticate) or validate it (check its health).
+
+            The C{prepareConnection} callable may raise an exception or return
+            a L{Deferred} which fails to reject the connection.  A rejected
```
Any particular exception type it's supposed to use to reject the connection "cleanly"?

glyph commented on src/twisted/application/internet.py:
```diff
@@ -1029,13 +1058,31 @@ def __init__(self, endpoint, factory, retryPolicy=None, clock=None):
+
+            The C{prepareConnection} callable may raise an exception or return
+            a L{Deferred} which fails to reject the connection.  A rejected
+            connection is not used to resolve an L{Deferred} returned by
```
Deferreds are generally said to be "fired" rather than "resolved" (which is the language one uses with JavaScript Promises). This isn't a dealbreaker, use this language if you prefer it, it's just a little inconsistent with the rest of Twisted.

glyph reviewed (approved):
Please address the one blocker here, documenting the version of the addition, and then I'd say this is approved to land.  (~~I don't see the little "member" flag next to your name, perhaps we should give you commit, but if you need a project member to land it, please just indicate this is pre-approved and doesn't need another review round.~~ looks like I was looking at the wrong place in the UI, land away :))

twm committed:
Merge branch 'trunk' into 8375-clientservice-onnewconnection

twm committed:
Update docstrings per review feedback

twm committed:
Correct epytext syntax

I can't find any way to use @since on a parameter, so text it is.

twm committed:
Merge branch '8375-clientservice-onnewconnection' of github.com:twisted/twisted into 8375-clientservice-onnewconnection

twm committed:
Fix random pyflakes errors

twm commented:
@glyph Thank you for the review! I have adjusted the terminology of the docstrings per your suggestion. I also submitted #1020 to better-document the syntax for `@since`. I will merge as soon as we have a solution to the Travis sendmsg test issues.

glyph commented:
@twm My understanding of the Travis issue is that it's unrelated. If it is actually caused by this change then by all means let's figure it out, of course.

Since it looked intermittent and unrelated to me though, I've re-run the builder and if it succeeds this time let's fix it separately.

glyph commented:
Bleh, I forgot that Travis doesn't save old logs before hitting the rebuild button. In any case, looks like that issue was unrelated, so please merge away.

twm commented:
@glyph I suppose the fourth time's the charm. :) I had given up mashing the button as it seemed to be failing consistently. Thanks again!

glyph commented:
When this sort of thing happens repeatedly on Travis, sometimes it's that we get unlucky and put onto a machine with a broken kernel or hardware configuration; weird APIs like `sendmsg` can be particularly sensitive to bad kernel versions, incorrect tuning, or overly strict container configuration.  Eventually they fix the issue and the intermittent failure stops.
